packages/store/addon/-private/identifiers/cache.ts

```diff
@@ -0,0 +1,426 @@
+import { DEBUG } from '@glimmer/env';
+import { assert } from '@ember/debug';
+import { Dict } from '../ts-interfaces/utils';
+import { ResourceIdentifierObject } from '../ts-interfaces/ember-data-json-api';
+import {
+  StableRecordIdentifier,
+  IS_IDENTIFIER,
+  DEBUG_CLIENT_ORIGINATED,
+  DEBUG_IDENTIFIER_BUCKET,
+  GenerationMethod,
+  UpdateMethod,
+  ForgetMethod,
+  ResetMethod,
+} from '../ts-interfaces/identifier';
+import coerceId from '../system/coerce-id';
+import uuidv4 from './utils/uuid-v4';
+import normalizeModelName from '../system/normalize-model-name';
+import isStableIdentifier from './is-stable-identifier';
+import isNonEmptyString from '../utils/is-non-empty-string';
+
+type IdentifierMap = Dict<string, StableRecordIdentifier>;
+type TypeMap = Dict<string, KeyOptions>;
+
+interface KeyOptions {
+  lid: IdentifierMap;
+  id: IdentifierMap;
+  _allIdentifiers: StableRecordIdentifier[];
+}
+
+let configuredForgetMethod: ForgetMethod;
+let configuredGenerationMethod: GenerationMethod;
+let configuredResetMethod: ResetMethod;
+let configuredUpdateMethod: UpdateMethod;
+
+export function setIdentifierGenerationMethod(method: GenerationMethod): void {
+  configuredGenerationMethod = method;
+}
+
+export function setIdentifierUpdateMethod(method: UpdateMethod): void {
+  configuredUpdateMethod = method;
+}
+
+export function setIdentifierForgetMethod(method: ForgetMethod): void {
+  configuredForgetMethod = method;
+}
+
+export function setIdentifierResetMethod(method: ResetMethod): void {
+  configuredResetMethod = method;
+}
+
+function defaultGenerationMethod(data: ResourceIdentifierObject, bucket: string): string {
+  if (isNonEmptyString(data.lid)) {
+    return data.lid;
+  }
+  let { type, id } = data;
+  if (isNonEmptyString(id)) {
+    return `@ember-data:lid-${normalizeModelName(type)}-${id}`;
```

Comment:
Does this work out with the edge cases of repeat usage and similar?

Reply:
it works out for most cases, individuals who need even stronger guarantees of uniqueness or whom already return an `id` unique across types can override the generation to a strategy that suits them better

Reply:
(this strategy being the default was discussed in the RFC fwiw)